ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DeprecatedStatementsChecker.java

```diff
@@ -0,0 +1,118 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.rest.server.computation;
+
+import static io.confluent.ksql.analyzer.Analysis.AliasedDataSource;
+
+import io.confluent.ksql.engine.rewrite.DataSourceExtractor;
+import io.confluent.ksql.metastore.MetaStore;
+import io.confluent.ksql.metastore.model.DataSource;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.Join;
+import io.confluent.ksql.parser.tree.JoinedSource;
+import io.confluent.ksql.parser.tree.Query;
+import io.confluent.ksql.parser.tree.Relation;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.parser.tree.WithinExpression;
+import java.util.Objects;
+import java.util.Optional;
+
+public class DeprecatedStatementsChecker {
+  public enum Deprecations {
+    DEPRECATED_STREAM_STREAM_JOIN_WITH_NO_GRACE(
+        "DEPRECATION NOTICE: Stream-stream joins statements without a GRACE PERIOD "
+            + "will not be accepted in a future ksqlDB version.\n"
+            + "Please use the GRACE PERIOD clause as specified in "
+            + "https://docs.ksqldb.io/en/latest/developer-guide/ksqldb-reference/"
+            + "select-push-query/"
+    );
+
+    private final String noticeMessage;
+
+    Deprecations(final String noticeMessage) {
+      this.noticeMessage = noticeMessage;
+    }
+
+    public String getNoticeMessage() {
+      return noticeMessage;
+    }
+  }
+
+  final MetaStore metaStore;
+
+  DeprecatedStatementsChecker(final MetaStore metaStore) {
+    this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+  }
+
+  public Optional<Deprecations> checkStatement(final Statement statement) {
+    if (isStreamStreamJoinWithoutGraceStatement(statement)) {
+      return Optional.of(Deprecations.DEPRECATED_STREAM_STREAM_JOIN_WITH_NO_GRACE);
+    }
+
+    return Optional.empty();
+  }
+
+  private boolean isStreamStreamJoinWithoutGraceStatement(final Statement statement) {
+    if (statement instanceof CreateAsSelect) {
+      return isStreamStreamJoinWithoutGraceQuery(((CreateAsSelect) statement).getQuery());
+    } else if (statement instanceof Query) {
+      return isStreamStreamJoinWithoutGraceQuery((Query) statement);
+    }
+
+    return false;
+  }
+
+  private boolean isStreamStreamJoinWithoutGraceQuery(final Query query) {
+    if (query.getFrom() instanceof Join) {
+      final Join join = (Join) query.getFrom();
+
+      // check left joined source is a stream
+      if (!isJoinedSourceStream(join.getLeft())) {
+        return false;
+      }
+
+      for (final JoinedSource joinedSource : join.getRights()) {
+        // check right joined source is a stream
+        if (!isJoinedSourceStream(joinedSource.getRelation())) {
+          break;
```

Comment:
Should this be `continue;` ? If we have a stream-table joins followed by a stream-stream join, the first `JoinedSource` in the list could not contain a STREAM but a TABLE as input, but we still need to check the second join, ie, third input, too?

Seems there is test for this case, but I think the test is setup incorrect (cf comment below)? Or I miss something?

Reply:
True. I switched to continue. The new test I added verifies this.